### PR TITLE
Implement Notifications for incoming tokens

### PR DESCRIPTION
**Pull request by justindg**
Note: Currently this only works when app is on foreground. The conditions for showing a notification can be found in `TransactionNotificationService.shouldShowNotification(tx, token)`.

A backend service that will help fetch transactions from background is currently in progress. After that is fully implemented, necessary blocks of code labeled with `TODO: [Notifications]` should be uncommented

### CODE REVIEW

File: app/src/main/java/com/alphawallet/app/ui/NotificationSettingsActivity.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+@AndroidEntryPoint
+public class NotificationSettingsActivity extends BaseActivity
+{
+    private NotificationSettingsViewModel viewModel;
+    private final ActivityResultLauncher<String> requestPermissionLauncher =
+        registerForActivityResult(new ActivityResultContracts.RequestPermission(), isGranted -> {
+            if (isGranted)
+            {
+                // FCM SDK (and your app) can post notifications.
+                viewModel.subscribe(1);
```
Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/ui/NotificationSettingsActivity.java:33]: 1 => MAINNET_ID

File: app/src/main/java/com/alphawallet/app/ui/NotificationSettingsActivity.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+//        {
+//            requestPostNotificationsPermission();
+//        }
+//        else
+//        {
+//            viewModel.unsubscribeToTopic(EthereumNetworkBase.MAINNET_ID);
+//        }
+    }
+
+    private void requestPostNotificationsPermission()
```
Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/ui/NotificationSettingsActivity.java:91]: Duplicated with `HomeActivity`.

Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/ui/NotificationSettingsActivity.java:91]: Duplicated with `HomeActivity`.

File: app/src/main/java/com/alphawallet/app/util/Utils.java [hunk 3]
```diff
@@ -556,6 +556,28 @@ public static String formatAddress(String address)
...
     }
 
+    public static String formatTxHash(String txHash)
+    {
+        if (isValidTxHash(txHash))
+        {
+            txHash = Keys.toChecksumAddress(txHash);
+            String result = "";
+            String firstSix = txHash.substring(0, 6);
+            String lastSix = txHash.substring(txHash.length() - 4);
```
Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/util/Utils.java:566]: Last four or last six?

File: app/src/main/java/com/alphawallet/app/viewmodel/HomeViewModel.java [hunk 4]
```diff
@@ -213,6 +218,7 @@ public void onClean()
 
     private void onDefaultWallet(final Wallet wallet)
     {
+        preferenceRepository.setWatchOnly(wallet.type == WalletType.WATCH);
```
Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/viewmodel/HomeViewModel.java:221]: `wallet.watchOnly()`

File: app/src/main/java/com/alphawallet/app/viewmodel/HomeViewModel.java [hunk 5]
```diff
@@ -829,4 +835,29 @@ private Callable<List<GitHubRelease>> getGitHubReleases(Request request)
             return null;
         };
     }
+
+    public void subscribeToNotifications()
+    {
+        alphaWalletNotificationService.subscribe(1);
```
Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/viewmodel/HomeViewModel.java:841]: 1 => MAINNET_ID

File: app/src/main/java/com/alphawallet/app/viewmodel/HomeViewModel.java [hunk 6]
```diff
@@ -829,4 +835,29 @@ private Callable<List<GitHubRelease>> getGitHubReleases(Request request)
...
+    }
+
+    public void unsubscribeToNotifications()
+    {
+        alphaWalletNotificationService.unsubscribeToTopic(1);
```
Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/viewmodel/HomeViewModel.java:846]: Same as above.

File: app/src/main/java/com/alphawallet/app/viewmodel/NotificationSettingsViewModel.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+            .subscribe(result -> Timber.d("unsubscribe result => " + result), Timber::e);
+    }
+
+    // TODO: [Notifications] Delete when unsubscribe is implemented
+    public void unsubscribeToTopic(long chainId)
+    {
+        alphaWalletNotificationService.unsubscribeToTopic(chainId);
+    }
+
+    public boolean getToggleState(String address)
```
Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/viewmodel/NotificationSettingsViewModel.java:79]: getToggleState => isTransactionNotificationsEnabled

File: app/src/main/java/com/alphawallet/app/viewmodel/NotificationSettingsViewModel.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    {
+        return preferenceRepository.isTransactionNotificationsEnabled(address);
+    }
+
+    public void setToggleState(String address, boolean state)
```
Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/viewmodel/NotificationSettingsViewModel.java:84]: state => enabled

File: app/src/main/java/com/alphawallet/app/viewmodel/SplashViewModel.java [hunk 9]
```diff
@@ -171,4 +171,12 @@ public void setInstallTime(long time)
     {
         preferenceRepository.setInstallTime(time);
     }
+
+    public void doWalletStartupActions(Wallet wallet)
+    {
+        boolean isWatchOnlyWallet = wallet.type == WalletType.WATCH;
```
Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/viewmodel/SplashViewModel.java:177]: wallet.watchOnly()

File: app/src/main/java/com/alphawallet/app/viewmodel/WalletActionsViewModel.java [hunk 10]
```diff
@@ -1,102 +1,133 @@
...
+    {
+        // TODO: Reactivate this when unsubscribe is implemented
+//        notificationDisposable =
+//            alphaWalletNotificationService.unsubscribe(1)
+//                .observeOn(Schedulers.io())
+//                .subscribeOn(Schedulers.io())
+//                .subscribe(result -> Timber.d("unsubscribe result => " + result), Timber::e);
+
+        // For now, unsubscribe to firebase topic
+        alphaWalletNotificationService.unsubscribeToTopic(1);
```
Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/viewmodel/WalletActionsViewModel.java:94]: 1 => MAINNET_ID

**Review by seabornlee — APPROVED** [review]

**Comment by seabornlee** [comment]
@justindg Could you please check the failed unit tests?

File: app/src/main/java/com/alphawallet/app/repository/SharedPreferenceRepository.java [hunk 11]
```diff
@@ -45,10 +46,12 @@ public class SharedPreferenceRepository implements PreferenceRepositoryType {
     private static final String SELECTED_SWAP_PROVIDERS_KEY = "selected_exchanges";
     private static final String ANALYTICS_KEY = "analytics_key";
     private static final String CRASH_REPORTING_KEY = "crash_reporting_key";
-
+    private static final String WALLET_CREATION_TIME = "wallet_creation_time";
```
[the remark this answers is not in the document]
Comment (justindg) [reply app/src/main/java/com/alphawallet/app/repository/SharedPreferenceRepository.java:49]: Nice catch. I reworked it, such that when the wallet will add a "creation time" if it doesn't exist yet. Also appended the address to the key, so multiple wallets will retain their own creation time.

**Review by JamesSmartCell — APPROVED** [review]
All works now, nice work

**Pull request merged**